### ISSUE TITLE
[1.10.x] Exception when default_factory uses function annotated with @deprecated from inherited model

### ISSUE THREAD

**Issue by lbenc135** [pydantic/pydantic#4925]
### Initial Checks

- [X] I have searched GitHub for a duplicate issue and I'm sure this is something new
- [X] I have searched Google & StackOverflow for a solution and couldn't find anything
- [X] I have read and followed [the docs](https://pydantic-docs.helpmanual.io) and still think this is a bug
- [X] I am confident that the issue is with pydantic (not my code, or another library in the ecosystem like [FastAPI](https://fastapi.tiangolo.com) or [mypy](https://mypy.readthedocs.io/en/stable))

### Description

The example provided below throws the following exception:
```
Traceback (most recent call last):
  File "/home/***/scratch.py", line 16, in <module>
    class Bar(Foo):
  File "pydantic/main.py", line 139, in pydantic.main.ModelMetaclass.__new__
  File "pydantic/utils.py", line 695, in pydantic.utils.smart_deepcopy
  File "/usr/lib/python3.10/copy.py", line 146, in deepcopy
    y = copier(x, memo)
  File "/usr/lib/python3.10/copy.py", line 231, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/usr/lib/python3.10/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/usr/lib/python3.10/copy.py", line 271, in _reconstruct
    state = deepcopy(state, memo)
  File "/usr/lib/python3.10/copy.py", line 146, in deepcopy
    y = copier(x, memo)
  File "/usr/lib/python3.10/copy.py", line 211, in _deepcopy_tuple
    y = [deepcopy(a, memo) for a in x]
  File "/usr/lib/python3.10/copy.py", line 211, in <listcomp>
    y = [deepcopy(a, memo) for a in x]
  File "/usr/lib/python3.10/copy.py", line 146, in deepcopy
    y = copier(x, memo)
  File "/usr/lib/python3.10/copy.py", line 231, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/usr/lib/python3.10/copy.py", line 153, in deepcopy
    y = copier(memo)
NotImplementedError: object proxy must define __deepcopy__()
```

The example is using this module: https://pypi.org/project/Deprecated/

By digging deeper I found that the `deprecated` module uses `wrapt.AdapterWrapper` which extends `wrapt.ObjectProxy` which then throws this exception if `__deepcopy__` is not overridden.

What is interesting is that there needs to be an inheritance (`Bar(Foo)`) for the exception to appear. I saw in Pydantic's code that there's some special handling of functions and other immutable types, so maybe we could add the same for wrapt's function wrapper.

### Example Code

```Python
import uuid

from deprecated import deprecated
from pydantic import BaseModel, Field


@deprecated
def random_id() -> str:
    return str(uuid.uuid4())


class Foo(BaseModel):
    id: str = Field(default_factory=random_id)


class Bar(Foo):
    pass
```

### Python, Pydantic & OS Version

```Text
'Deprecated' version: 1.2.13

pydantic version: 1.10.2
pydantic compiled: True
install path: /home/***/venv/lib/python3.10/site-packages/pydantic
               python version: 3.10.6 (main, Nov 14 2022, 16:10:14) [GCC 11.3.0]
                     platform: Linux-5.15.0-57-generic-x86_64-with-glibc2.35
     optional deps. installed: ['dotenv', 'email-validator', 'typing-extensions']
```

### Affected Components

- [ ] [Compatibility between releases](https://pydantic-docs.helpmanual.io/changelog/)
- [ ] [Data validation/parsing](https://pydantic-docs.helpmanual.io/usage/models/#basic-model-usage)
- [ ] [Data serialization](https://pydantic-docs.helpmanual.io/usage/exporting_models/) - `.model_dump()` and `.model_dump_json()`
- [ ] [JSON Schema](https://pydantic-docs.helpmanual.io/usage/schema/)
- [ ] [Dataclasses](https://pydantic-docs.helpmanual.io/usage/dataclasses/)
- [ ] [Model Config](https://pydantic-docs.helpmanual.io/usage/model_config/)
- [ ] [Field Types](https://pydantic-docs.helpmanual.io/usage/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://pydantic-docs.helpmanual.io/usage/validation_decorator/)
- [ ] [Generic Models](https://pydantic-docs.helpmanual.io/usage/models/#generic-models)
- [X] [Other Model behaviour](https://pydantic-docs.helpmanual.io/usage/models/) - `model_construct()`, pickling, private attributes, ORM mode
- [ ] [Plugins](https://pydantic-docs.helpmanual.io/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

**Comment by hramezani** [comment]
Thanks @lbenc135 for reporting!

I can confirm the problem is reproducible on `1.10.x` but I couldn't reproduce it on `main`.

As we are working on `V2` and the plan is to just provide bug fix patches for `1.10.x`, I think we are not going to fix it on `1.10.x` and it will be fixed in `V2`.

I think we can close this bug @samuelcolvin

**Comment by samuelcolvin** [comment]
If there's an easy fix, happy to consider it for a v1.10.X patch, but as @hramezani says, our priority is currently V2.

I'll leave this open for now.
